custom_components/egauge/sensor.py

```diff
@@ -62,54 +64,75 @@ class EGaugeSensor(EGaugeEntity, SensorEntity):
 
     def __init__(
         self,
-        data_type,
+        data_type: str,
         register_name,
         register_type_code,
         interval,
         coordinator,
         config_entry,
-    ):
+    ) -> None:
         self.data_type = data_type
         self.register_name = register_name
         self.register_type_code = register_type_code
         self.interval = interval
         self.unit_conversion = EGAUGE_UNIT_CONVERSIONS[self.data_type].get(
             self.register_type_code, 1.0
         )
+
+        _LOGGER.info(
+            "register_name: %s  type: %s  register_type_code: %s  interval: %s  "
+            "unit_conversion: %s",
+            self.register_name,
+            self.data_type,
+            self.register_type_code,
+            self.interval,
+            self.unit_conversion,
+        )
+
         super().__init__(coordinator, config_entry)
 
     @property
-    def is_historical(self):
+    def is_historical(self) -> bool:
+        """True if this sensor represents a historical value, false otherwise."""
         return self.data_type == EGAUGE_HISTORICAL
 
     @property
-    def unique_id(self):
+    def unique_id(self) -> str:
         """Return a unique ID to use for this entity."""
         if self.is_historical:
             return f"{self.entry_id}-{self.interval}-{self.register_name}"
-        else:
-            return f"{self.entry_id}-{self.register_name}"
+        return f"{self.entry_id}-{self.register_name}"
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Return the name of the sensor."""
         if self.is_historical:
             return f"{DEFAULT_NAME} {self.interval} {self.register_name}"
-        else:
-            return f"{DEFAULT_NAME} {self.register_name}"
+        return f"{DEFAULT_NAME} {self.register_name}"
 
     @property
-    def state(self):
+    def state(self) -> str:
         """Return the state of the sensor."""
         data = self.coordinator.data[self.data_type]
         if self.is_historical:
             data = data[self.interval]
         value = data.get(self.register_name)
         value = value * self.unit_conversion
+        if self.state_class == SensorStateClass.TOTAL_INCREASING and value < 0:
+            pv = value
+            value = abs(value)
```

Comment:
From what I've seen, I'd guess this happens due to rounding errors somewhere along the line. I think it's probably better to clamp this to 0 rather than taking the absolute value.

Reply:
After looking at this more closely, I'm not sure it's just rounding. I extracted some log lines for one sensor:

```
2024-11-21 12:59:33.978 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.671942222222222); setting to zero
2024-11-21 13:00:04.965 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.673531944444444); setting to zero
2024-11-21 13:00:36.068 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6751116666666666); setting to zero
2024-11-21 13:01:07.220 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.676646111111111); setting to zero
2024-11-21 13:01:38.002 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6781794444444444); setting to zero
2024-11-21 13:02:09.034 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6797402777777775); setting to zero
2024-11-21 13:02:40.043 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.681293333333333); setting to zero
2024-11-21 13:03:11.030 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.682842222222222); setting to zero
2024-11-21 13:03:41.876 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.684393333333333); setting to zero
2024-11-21 13:04:13.087 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6859355555555555); setting to zero
2024-11-21 13:04:44.191 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6874683333333333); setting to zero
2024-11-21 13:05:14.860 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6890052777777775); setting to zero
2024-11-21 13:05:46.293 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.690536388888889); setting to zero
2024-11-21 13:06:17.052 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6920719444444443); setting to zero
2024-11-21 13:06:47.995 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6935955555555555); setting to zero
2024-11-21 13:07:19.301 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6951199999999997); setting to zero
2024-11-21 13:07:50.075 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6966341666666667); setting to zero
2024-11-21 13:08:20.957 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 1 is total_increasing yet value is negative (-2.6981413888888888); setting to zero
```

Note that the values seem to be constantly decreasing.

Here's another sensor:

```
2024-11-21 12:58:01.039 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.317182222222222); setting to zero
2024-11-21 12:58:31.881 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3191599999999997); setting to zero
2024-11-21 12:59:03.188 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3211375); setting to zero
2024-11-21 12:59:33.978 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3231213888888886); setting to zero
2024-11-21 13:00:04.964 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.325094722222222); setting to zero
2024-11-21 13:00:36.067 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.327125833333333); setting to zero
2024-11-21 13:01:07.220 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3291275); setting to zero
2024-11-21 13:01:38.002 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3314552777777777); setting to zero
2024-11-21 13:02:09.033 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3336155555555553); setting to zero
2024-11-21 13:02:40.042 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.335590555555555); setting to zero
2024-11-21 13:03:11.022 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.33756); setting to zero
2024-11-21 13:03:41.876 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3395444444444444); setting to zero
2024-11-21 13:04:13.086 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3415130555555552); setting to zero
2024-11-21 13:04:44.190 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3434902777777777); setting to zero
2024-11-21 13:05:14.860 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.345468055555555); setting to zero
2024-11-21 13:05:46.293 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.347447222222222); setting to zero
2024-11-21 13:06:17.050 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3494313888888887); setting to zero
2024-11-21 13:06:47.995 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3514055555555555); setting to zero
2024-11-21 13:07:19.300 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3533927777777777); setting to zero
2024-11-21 13:07:50.075 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.355361111111111); setting to zero
2024-11-21 13:08:20.956 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3573302777777774); setting to zero
2024-11-21 13:08:51.956 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.359294722222222); setting to zero
2024-11-21 13:09:22.988 DEBUG (MainThread) [custom_components.egauge] egauge todays Sensor 2 is total_increasing yet value is negative (-3.3612636111111107); setting to zero
```

These are also constantly decreasing. Seems like rounding errors wouldn't be cumulative?

Reply:
Okay, I see. I think you're getting this for a different reason than how I see it on my setup. I see it every once in a while around the time the sensor resets. It looks like you have a sensor in egauge that just always returns negative values.

I think the right fix for this might actually be more involved. We should switch the sensor type from `TOTAL_INCREASING` to `TOTAL` (as describe [here](https://developers.home-assistant.io/blog/2021/09/20/state_class_total/)). I need to research what the effects of changing a sensor type would be on existing installations, though.

For now, do you mind just removing the fix altogether while I figure out the right solution?